## Diabetes Challenge

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 60)
import warnings
warnings.filterwarnings('ignore')

Install the repo package and import the data

In [2]:
!pip install ucimlrepo

In [3]:
from ucimlrepo import fetch_ucirepo
diabetes = fetch_ucirepo(id=296)
diabetes.data.original.head(5)

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)    NaN   
1        149190     55629189        Caucasian  Female  [10-20)    NaN   
2         64410     86047875  AfricanAmerican  Female  [20-30)    NaN   
3        500364     82442376        Caucasian    Male  [30-40)    NaN   
4         16680     42519267        Caucasian    Male  [40-50)    NaN   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital payer_code         medical_specialty  num_lab_procedures  \
0                 1        NaN  Pediatrics-Endocrinology                  41   
1                 3        NaN                       NaN                  59   
2                 2        NaN                       NaN                  11   
3                 2        NaN                       NaN                  44   
4                 1        NaN                       NaN                  51   

   num_procedures  num_medications  number_outpatient  number_emergency  \
0               0                1                  0                 0   
1               0               18                  0                 0   
2               5               13                  2                 0   
3               1               16                  0                 0   
4               0                8                  0                 0   

   number_inpatient  diag_1  diag_2 diag_3  number_diagnoses max_glu_serum  \
0                 0  250.83     NaN    NaN                 1           NaN   
1                 0     276  250.01    255                 9           NaN   
2                 1     648     250    V27                 6           NaN   
3                 0       8  250.43    403                 7           NaN   
4                 0     197     157    250                 5           NaN   

  A1Cresult metformin repaglinide nateglinide chlorpropamide glimepiride  \
0       NaN        No          No          No             No          No   
1       NaN        No          No          No             No          No   
2       NaN        No          No          No             No          No   
3       NaN        No          No          No             No          No   
4       NaN        No          No          No             No          No   

  acetohexamide glipizide glyburide tolbutamide pioglitazone rosiglitazone  \
0            No        No        No          No           No            No   
1            No        No        No          No           No            No   
2            No    Steady        No          No           No            No   
3            No        No        No          No           No            No   
4            No    Steady        No          No           No            No   

  acarbose miglitol troglitazone tolazamide examide citoglipton insulin  \
0       No       No           No         No      No          No      No   
1       No       No           No         No      No          No      Up   
2       No       No           No         No      No          No      No   
3       No       No           No         No      No          No      Up   
4       No       No           No         No      No          No  Steady   

  glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
0                  No                  No                       No   
1                  No                  No                       No   
2                  No                  No                       No   
3                  No                  No                       No   
4        

The source of the data is found in: https://archive.ics.uci.edu/dataset/296/diabetes+130-us+hospitals+for+years+1999-2008.

A paper was published in 2014 at the BioMed Research International from the Wiley editorial. En este paper se evalua este mismo dataset y usando una regresión logistica multivariable se encuentra una relación entre las readmisiones y la medición del HbA1c. El articulo concluye que las tasas de reingreso para los pacientes con diabetes parecen estar más asociadas a la decisión de hacer la medición del HbA1c, más que con los resultados.



En la fuente de la data se encuentra un resumen de cada variable, su tipo de dato y si tiene faltantes. Adicionalmente, en el paper encontramos la misma información aclarando la cantidad de data faltante y la interpretación de cada variable.

Inicialmente, vamos a validar nuestra información y contrastarla con los datos presentados en el paper

In [4]:
#Tamaño de la muestra
len(diabetes.data.original)

101766

Al igual que el paper se tienen 101766 encuentros

In [5]:
#Tipos de variables
diabetes.variables

name     role         type demographic  \
0               encounter_id       ID                     None   
1                patient_nbr       ID                     None   
2                       race  Feature  Categorical        Race   
3                     gender  Feature  Categorical      Gender   
4                        age  Feature  Categorical         Age   
5                     weight  Feature  Categorical        None   
6          admission_type_id  Feature  Categorical        None   
7   discharge_disposition_id  Feature  Categorical        None   
8        admission_source_id  Feature  Categorical        None   
9           time_in_hospital  Feature      Integer        None   
10                payer_code  Feature  Categorical        None   
11         medical_specialty  Feature  Categorical        None   
12        num_lab_procedures  Feature      Integer        None   
13            num_procedures  Feature      Integer        None   
14           num_medications  Feature      Integer        None   
15         number_outpatient  Feature      Integer        None   
16          number_emergency  Feature      Integer        None   
17          number_inpatient  Feature      Integer        None   
18                    diag_1  Feature  Categorical        None   
19                    diag_2  Feature  Categorical        None   
20                    diag_3  Feature  Categorical        None   
21          number_diagnoses  Feature      Integer        None   
22             max_glu_serum  Feature  Categorical        None   
23                 A1Cresult  Feature  Categorical        None   
24                 metformin  Feature  Categorical        None   
25               repaglinide  Feature  Categorical        None   
26               nateglinide  Feature  Categorical        None   
27            chlorpropamide  Feature  Categorical        None   
28               glimepiride  Feature  Categorical        None   
29             acetohexamide  Feature  Categorical        None   
30                 glipizide  Feature  Categorical        None   
31                 glyburide  Feature  Categorical        None   
32               tolbutamide  Feature  Categorical        None   
33              pioglitazone  Feature  Categorical        None   
34             rosiglitazone  Feature  Categorical        None   
35                  acarbose  Feature  Categorical        None   
36                  miglitol  Feature  Categorical        None   
37              troglitazone  Feature  Categorical        None   
38                tolazamide  Feature  Categorical        None   
39                   examide  Feature  Categorical        None   
40               citoglipton  Feature  Categorical        None   
41                   insulin  Feature  Categorical        None   
42       glyburide-metformin  Feature  Categorical        None   
43       glipizide-metformin  Feature  Categorical        None   
44  glimepiride-pioglitazone  Feature  Categorical        None   
45   metformin-rosiglitazone  Feature  Categorical        None   
46    metformin-pioglitazone  Feature  Categorical        None   
47                    change  Feature  Categorical        None   
48               diabetesMed  Feature  Categorical        None   
49                readmitted   Target  Categorical        None   

                                          description units missing_values  
0                   Unique identifier of an encounter  None             no  
1                      Unique identifier of a patient  None             no  
2   Values: Caucasian, Asian, African American, Hi...  None            yes  
3           Values: male, female, and unknown/invalid  None             no  
4   Grouped in 10-year intervals: [0, 10), [10, 20...  None             no  
5                                   Weight in pounds.  None            yes  
6   Integer identifier corresponding to 9 distinct...  None             no  
7   Integer identifier corresponding to 2

In [6]:
#Data faltante
diabetes.data.original.isna().sum()/len(diabetes.data.original) *100

encounter_id                 0.000000
patient_nbr                  0.000000
race                         2.233555
gender                       0.000000
age                          0.000000
weight                      96.858479
admission_type_id            0.000000
discharge_disposition_id     0.000000
admission_source_id          0.000000
time_in_hospital             0.000000
payer_code                  39.557416
medical_specialty           49.082208
num_lab_procedures           0.000000
num_procedures               0.000000
num_medications              0.000000
number_outpatient            0.000000
number_emergency             0.000000
number_inpatient             0.000000
diag_1                       0.020636
diag_2                       0.351787
diag_3                       1.398306
number_diagnoses             0.000000
max_glu_serum               94.746772
A1Cresult                   83.277322
metformin                    0.000000
repaglinide                  0.000000
nateglinide                  0.000000
chlorpropamide               0.000000
glimepiride                  0.000000
acetohexamide                0.000000
glipizide                    0.000000
glyburide                    0.000000
tolbutamide                  0.000000
pioglitazone                 0.000000
rosiglitazone                0.000000
acarbose                     0.000000
miglitol                     0.000000
troglitazone                 0.000000
tolazamide                   0.000000
examide                      0.000000
citoglipton                  0.000000
insulin                      0.000000
glyburide-metformin          0.000000
glipizide-metformin          0.000000
glimepiride-pioglitazone     0.000000
metformin-rosiglitazone      0.000000
metformin-pioglitazone       0.000000
change                       0.000000
diabetesMed                  0.000000
readmitted                   0.000000
dtype: float64

Al comparar con la data faltante que reporta el paper encontramos los mismos valores

Con la información vista hasta ahora y la del paper vamos a realizar un proceso de limpieza: Inicialmente sacamos las columnas de **weight** y **payer_code**. En busca de una comparación futura mantenemos la columna **medical_specialty** y realizamos la misma transformación. Adicionalmente, vamos a cambiar rellenar la información con none en los casos donde falte.

In [7]:
diabetes.data.original = diabetes.data.original.drop(columns=['weight', 'payer_code']) #Eliminamos columnas
diabetes.data.original["medical_specialty"].fillna("missing", inplace = True) #Transformamos columna
diabetes.data.original.fillna("none", inplace = True)

En el articulo utilizan solo la primera visita del paciente para garantizar independencia estadistica, lo cual es necesario para la regresión logistica usada por Strack, et al.

En este punto podemos tener dos acercamientos. En el primero vamos a garantarizar la independencia estadistica que propone Strack. En el segundo vamos a tomar todos los casos (incluyendo revisitas) con la salvedad de que se utilizaran modelos donde no importe la dependencia entre ejemplos como lo serían como lo son:
* Algoritmos de ensamble
* Redes neuronales
* K-means (Clusterización)

entre otros.

### Data con independencia estadistica (Strack's data)

En este punto se hará un proceso de recrear la limpieza llevada a cabo por Strack.

Inicialmente nos quedamos solo con las primeras visitas de cada paciente.

In [8]:
Strack_Diabetes = diabetes.data.original
Strack_Diabetes = Strack_Diabetes.sort_values(by=['patient_nbr', 'encounter_id' ])
Strack_Diabetes = Strack_Diabetes.drop_duplicates(subset=['patient_nbr'], keep='first')
len(Strack_Diabetes)

71518

Ahora quitamos los pacientes que murieron o fueron trasladados a cuidados paleativos. Estos dos casos corresponden a los ids 11 y 13, 14 (respectivamente). Esta información se extrae del archivo de IDS_mapping donde explican el significado de cada ID

In [9]:
Strack_Diabetes = Strack_Diabetes[~Strack_Diabetes['discharge_disposition_id'].isin([11, 13, 14])]
len(Strack_Diabetes)

69980

En este punto ya tenemos la data tal cual se presenta en el articulo de Strack. Esta data la vamos a guardar

In [10]:
Strack_Diabetes.to_csv('Strack_Diabetes.csv')

### Data con dependencia estadistica

In [14]:
Hernandez_Diabetes = diabetes.data.original
Hernandez_Diabetes = Hernandez_Diabetes.sort_values(by=['patient_nbr', 'encounter_id' ])
Hernandez_Diabetes.head(10)

encounter_id  patient_nbr             race  gender      age  \
4267       24437208          135        Caucasian  Female  [50-60)   
4780       26264286          135        Caucasian  Female  [50-60)   
5827       29758806          378        Caucasian  Female  [50-60)   
67608     189899286          729        Caucasian  Female  [80-90)   
17494      64331490          774        Caucasian  Female  [80-90)   
2270       14824206          927  AfricanAmerican  Female  [30-40)   
1164        8380170         1152  AfricanAmerican  Female  [50-60)   
5953       30180318         1152  AfricanAmerican  Female  [50-60)   
14180      55533660         1152  AfricanAmerican  Female  [60-70)   
23623      80742510         1152  AfricanAmerican  Female  [60-70)   

       admission_type_id  discharge_disposition_id  admission_source_id  \
4267                   2                         1                    1   
4780                   1                         1                    7   
5827                   3                         1                    1   
67608                  1                         3                    7   
17494                  1                         1                    7   
2270                   1                         1                    7   
1164                   1                         1                    7   
5953                   1                         1                    7   
14180                  1                         1                    7   
23623                  1                         1                    7   

       time_in_hospital                medical_specialty  num_lab_procedures  \
4267                  8                       Cardiology                  77   
4780                  3  Surgery-Cardiovascular/Thoracic                  31   
5827                  2                    Surgery-Neuro                  49   
67608                 4                 InternalMedicine                  68   
17494                 3                 InternalMedicine                  46   
2270                  5                 InternalMedicine                  49   
1164                  6              Hematology/Oncology                  43   
5953                  6              Hematology/Oncology                  45   
14180                10              Hematology/Oncology                  54   
23623                 8                          missing                  30   

       num_procedures  num_medications  number_outpatient  number_emergency  \
4267                6               33                  0                 0   
4780                1               14                  0                 0   
5827                1               11                  0                 0   
67608               2               23                  0                 0   
17494               0               20                  0                 0   
2270                0                5                  0                 0   
1164                2               13                  0                 0   
5953                4               15                  0                 0   
14180               2               19                  0                 0   
23623               1               16                  0                 0   

       number_inpatient diag_1  diag_2 diag_3  number_diagnoses max_glu_serum  \
4267                  0    401     997    560                 8          none   
4780                  1    998      41    250                 5          none   
5827                  0    722     305    250                 3          none   
67608                 0    820     493   E880                 9          none   
17494                 0    274     427    416                 9          none   
2270                  0    590     220    250                 3          none   
1164                  1    282  250.01   none                 2          none   
5953       

La gran diferencia es que no vamos a quitar las revisitas. Por el contrario vamos a crear una nueva columna que nos indique cual visita es de cada paciente, la primera, segunda, etc. Esta columna a pesar de ser parecida a number_inpatient se diferencia en el tiempo en que se cuenta una revisita.

In [16]:
Hernandez_Diabetes['all_time_number_inpatient'] = Hernandez_Diabetes.groupby('patient_nbr').cumcount() + 1

Ahora vamos a revisar la otra limpieza que se hizo para el dataset de Strack donde quitamos muertos y traslados a cuidados paleativos

In [25]:
Death = Hernandez_Diabetes[Hernandez_Diabetes['discharge_disposition_id'] == 11]
Death.head(2)

encounter_id  patient_nbr       race  gender      age  \
8102      37264596        61011  Caucasian  Female  [40-50)   
9869      42430920        80046  Caucasian    Male  [80-90)   

      admission_type_id  discharge_disposition_id  admission_source_id  \
8102                  1                        11                    7   
9869                  1                        11                    7   

      time_in_hospital medical_specialty  num_lab_procedures  num_procedures  \
8102                 3           missing                  53               0   
9869                10  InternalMedicine                  70               1   

      num_medications  number_outpatient  number_emergency  number_inpatient  \
8102               17                  0                 0                 0   
9869               22                  0                 0                 0   

     diag_1 diag_2 diag_3  number_diagnoses max_glu_serum A1Cresult metformin  \
8102    599    452    305                 5          none      none        No   
9869    428    427    425                 9          none      none        No   

     repaglinide nateglinide chlorpropamide glimepiride acetohexamide  \
8102          No          No             No          No            No   
9869          No          No             No          Up            No   

     glipizide glyburide tolbutamide pioglitazone rosiglitazone acarbose  \
8102        No        No          No           No            No       No   
9869        No        No          No           No            No       No   

     miglitol troglitazone tolazamide examide citoglipton insulin  \
8102       No           No         No      No          No      No   
9869       No           No         No      No          No      No   

     glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
8102                  No                  No                       No   
9869                  No                  No                       No   

     metformin-rosiglitazone metformin-pioglitazone change diabetesMed  \
8102                      No                     No     No          No   
9869                      No                     No     Ch         Yes   

     readmitted  all_time_number_inpatient  
8102         NO                          1  
9869         NO                          1

In [29]:
Death['patient_nbr'].value_counts()

patient_nbr
61011        1
74462598     1
75551022     1
75489480     1
75196629     1
            ..
25900407     1
25848126     1
25731189     1
25729560     1
189179321    1
Name: count, Length: 1642, dtype: int64

In [28]:
Death['readmitted'].value_counts()

readmitted
NO    1642
Name: count, dtype: int64

Como se esperaba todas las personas muertas tienen el label de NO en readmitidos. Esta prueba sencilla es una muestra de que la data se encuentra bien catalogada.

Esta base puede servir para identificar un patrón entre las personas fallecidas. El NO en el readmitted puede significar que las personas se curan o mueren. En este sentido tiene sentido sacarlos de la data Hernandez_Diabetes. Pero dejarlos en otra base para posterior estudio. Otro posible resultado es crear una nueva categoria para los muertos.

Vamos a remover muertos y dejarlos en una base independiente para posterior análisis

In [31]:
Death.to_csv('Death.csv')
Hernandez_Diabetes = Hernandez_Diabetes[Hernandez_Diabetes['discharge_disposition_id'] != 11]
Hernandez_Diabetes.head(2)

encounter_id  patient_nbr       race  gender      age  \
4267      24437208          135  Caucasian  Female  [50-60)   
4780      26264286          135  Caucasian  Female  [50-60)   

      admission_type_id  discharge_disposition_id  admission_source_id  \
4267                  2                         1                    1   
4780                  1                         1                    7   

      time_in_hospital                medical_specialty  num_lab_procedures  \
4267                 8                       Cardiology                  77   
4780                 3  Surgery-Cardiovascular/Thoracic                  31   

      num_procedures  num_medications  number_outpatient  number_emergency  \
4267               6               33                  0                 0   
4780               1               14                  0                 0   

      number_inpatient diag_1 diag_2 diag_3  number_diagnoses max_glu_serum  \
4267                 0    401    997    560                 8          none   
4780                 1    998     41    250                 5          none   

     A1Cresult metformin repaglinide nateglinide chlorpropamide glimepiride  \
4267      none    Steady          No          No             No          No   
4780      none    Steady          No          No             No          No   

     acetohexamide glipizide glyburide tolbutamide pioglitazone rosiglitazone  \
4267            No        No      Down          No           No            No   
4780            No        No    Steady          No           No            No   

     acarbose miglitol troglitazone tolazamide examide citoglipton insulin  \
4267       No       No           No         No      No          No  Steady   
4780       No       No           No         No      No          No      No   

     glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
4267                  No                  No                       No   
4780                  No                  No                       No   

     metformin-rosiglitazone metformin-pioglitazone change diabetesMed  \
4267                      No                     No     Ch         Yes   
4780                      No                     No     Ch         Yes   

     readmitted  all_time_number_inpatient  
4267        <30                          1  
4780        >30                          2

A pesar de que el target es **readmitted**, este resultado tiene más importancia al sacar a los casos muertos. Adicionalmente, se debe realizar un análisis cuidadoso de la columna **discharge_disposition_id**.

In [32]:
import plotly.express as px
#vamos a graficar en el eje x readmitted y en el eje y el numero de data que se tiene para cada discharge_dispotition_id
fig = px.histogram(Hernandez_Diabetes,
                   x='readmitted',
                   color='discharge_disposition_id',
                   barmode='stack',
                   title='Cantidad de readmitted para discharge_disposition_id',
                   labels={'readmitted': 'readmitted', 'count': 'discharge_disposition_id count'})

# Mostrar el gráfico
fig.show()

Al observar cada discharge_disposition_id observamos que los id 19 y 20 tiene todos sus valores de **readmitted** en NO. Ambos ID corresponden a muerte en casa o en medical facility. Entre ambos Ids tenemos 10 casos. Por su significado y la poca cantidad de data que representan se omitiran de la data Hernandez_Diabetes

In [34]:
Hernandez_Diabetes = Hernandez_Diabetes[~Hernandez_Diabetes['discharge_disposition_id'].isin([19, 20])]

Con esto concluiriamos la limpieza de la data y tendriamos una segunda base con mayor cantidad de data

In [36]:
len(Hernandez_Diabetes)

100114

In [35]:
Hernandez_Diabetes.to_csv('Hernandez_Diabetes.csv')

### Palabras finales

En este notebook se crearon 3 datasets diferentes.

El primero corresponde al dataset usado por Strack en su articulo *http://dx.doi.org/10.1155/2014/781670*. En este dataset se garantiza la independencia estadistica al quedarnos solo con la primera visita de cada paciente. Adicionalmente se eliminan pacientes muertos y que entran en cuidados paleativos

La segunda base corresponde al dataset creado por el autor. En este dataset se tienen en cuenta las multiples visitas que puede tener un paciente y se enumeran. Despues de analizar el target se extraen del dataset los muertos.

La tercera base corresponde a los muertos que fueron excluidos en el dataset Hernandez_Diabetes. Este dataset se utilizará para buscar información relevante y patrones de las personas fallecidas.